runtime/bigarray.c

```diff
@@ -98,22 +98,23 @@ caml_ba_alloc(int flags, int num_dims, void * data, intnat * dim)
   CAMLassert((flags & CAML_BA_KIND_MASK) <= CAML_BA_CHAR);
   for (i = 0; i < num_dims; i++) dimcopy[i] = dim[i];
   size = 0;
-  if (data == NULL) {
-    num_elts = 1;
-    for (i = 0; i < num_dims; i++) {
-      if (caml_umul_overflow(num_elts, dimcopy[i], &num_elts))
-        caml_raise_out_of_memory();
-    }
-    if (caml_umul_overflow(num_elts,
-                           caml_ba_element_size[flags & CAML_BA_KIND_MASK],
-                           &size))
+  num_elts = 1;
+  for (i = 0; i < num_dims; i++) {
+    if (caml_umul_overflow(num_elts, dimcopy[i], &num_elts))
       caml_raise_out_of_memory();
+  }
+  if (caml_umul_overflow(num_elts,
+                         caml_ba_element_size[flags & CAML_BA_KIND_MASK],
+                         &size))
+    caml_raise_out_of_memory();
+  if (data == NULL) {
     data = malloc(size);
     if (data == NULL && size != 0) caml_raise_out_of_memory();
     flags |= CAML_BA_MANAGED;
   }
   asize = SIZEOF_BA_ARRAY + num_dims * sizeof(intnat);
-  res = caml_alloc_custom_mem(&caml_ba_ops, asize, size);
+  res = caml_alloc_custom_mem(&caml_ba_ops, asize,
+                              (flags & CAML_BA_MANAGED) ? size : 0);
```

Comment:
You're supposed to mask with `CAML_BA_MANAGED_MASK` and compare with `CAML_BA_EXTERNAL` or `CAML_BA_MANAGED`. (compare `caml_ba_finalize` around line 146)

```suggestion
  if ((flags & CAML_BA_MANAGED_MASK) == CAML_BA_EXTERNAL) size = 0;
  res = caml_alloc_custom_mem(&caml_ba_ops, asize, size);
```

Reply:
Thanks, fixed. (I also removed a redundant initialisation a couple of lines up)